File: cloud_frontend/azure/python/storage.py

import os
import uuid

from azure.storage.blob import BlobServiceClient


class storage:
    instance = None
    client = None

    def __init__(self):
        self.client = BlobServiceClient.from_connection_string(
                os.getenv('STORAGE_CONNECTION_STRING')
            )

    @staticmethod
    def unique_name(name):
        name, extension = name.split('.')
        return '{name}.{random}.{extension}'.format(
                    name=name,
                    extension=extension,
                    random=str(uuid.uuid4()).split('-')[0]
                )

    def upload(self, container, file, filepath):
        with open(filepath, 'rb') as data:
            self.upload_stream(container, storage.unique_name(file), data)

    def download(self, container, file, filepath):
        with open(filepath, 'wb') as download_file:
            download_file.write( self.download_stream(container, file) )
    
    def download_directory(self, container, prefix, path):
        client = self.client.get_container_client(container=container)
        objects = client.list_blobs(name_starts_with=prefix)
        for obj in objects:
            file_name = obj.name
            path_to_file = os.path.dirname(file_name)
            os.makedirs(os.path.join(path, path_to_file), exist_ok=True)
            self.download(container, file_name, os.path.join(path, file_name))
    
    # it seems that JS does not have an API that would allow to 
    # upload/download data without going through container client
    def upload_stream(self, container, file, data):
        client = self.client.get_blob_client(
                container=container,
                blob=storage.unique_name(file)
        )
        return client.upload_blob(data)

    def download_stream(self, container, file):
        client = self.client.get_blob_client(container=container, blob=file)
        return client.download_blob().readall()
    
    def get_instance():
        if storage.instance is None:
            storage.instance = storage()
        return storage.instance

File: cloud_frontend/azure/nodejs/storage.js

const { BlobServiceClient } = require('@azure/storage-blob'),
        stream = require('stream');


class azure_storage {

  constructor() {
    const STORAGE_CONNECTION_STRING = process.env.STORAGE_CONNECTION_STRING;   
    this.client = BlobServiceClient.fromConnectionString(STORAGE_CONNECTION_STRING);
  }

  upload(bucket, file, filepath) {
    // TOD:
  };

  download(bucket, file, filepath) {
    // TODO:
  };

  // We could provide additional API for just providing a byte buffer and uploading
  // Right now streams seems to be nicer
  uploadStream(container, file) {
    // it seems that JS does not have an API that would allow to
    // upload/download data without going through container/blob client 
    let containerClient = this.client.getContainerClient(container);
    let blockBlobClient = containerClient.getBlockBlobClient(file);
    var write_stream = new stream.PassThrough();
    // returns promise
    let upload = blockBlobClient.uploadStream(write_stream);
    return [write_stream, upload];
  };

  downloadStream(container, file) {
    let containerClient = this.client.getContainerClient(container);
    let blockBlobClient = containerClient.getBlockBlobClient(file);
    return blockBlobClient.download().then(
        (data) => {
          return data.readableStreamBody;
        }
    );
  };
};
exports.storage = azure_storage;
